validation/src/header.rs

```diff
@@ -208,19 +215,21 @@ fn compute_next_difficulty(
     // returned Regtest network doesn't adjust PoW difficult levels. For
     // regtest, simply return the previous difficulty target
 
-    if (prev_height + 1) % DIFFICULTY_ADJUSTMENT_INTERVAL != 0 || no_pow_retargeting(network) {
+    let height = prev_height + 1;
+    if height % DIFFICULTY_ADJUSTMENT_INTERVAL != 0 || no_pow_retargeting(network) {
         return prev_header.bits;
     }
 
-    // Computing the last header with height multiple of 2016
-    let mut current_header = *prev_header;
-    for _i in 0..(DIFFICULTY_ADJUSTMENT_INTERVAL - 1) {
-        if let Some((header, _)) = store.get_header(&current_header.prev_blockhash) {
-            current_header = header;
-        }
-    }
-    // last_adjustment_header is the last header with height multiple of 2016
-    let last_adjustment_header = current_header;
+    // Computing the `last_adjustment_header`.
+    // `last_adjustment_header` is the last header with height multiple of 2016
+    let last_adjustment_height = if height < DIFFICULTY_ADJUSTMENT_INTERVAL {
+        0
+    } else {
+        height - DIFFICULTY_ADJUSTMENT_INTERVAL
+    };
```

Comment:
This is an optimization from the previous code. Rather than backtracking 2016 blocks, we simply fetch that block directly.